node/heartbeat/hearbeatMessageInfo.go

```diff
@@ -22,81 +20,101 @@ type heartbeatMessageInfo struct {
 	nodeDisplayName    string
 	isValidator        bool
 	lastUptimeDowntime time.Time
+	genesisTime        time.Time
 }
 
 // newHeartbeatMessageInfo returns a new instance of a heartbeatMessageInfo
 func newHeartbeatMessageInfo(
 	maxDurationPeerUnresponsive time.Duration,
 	isValidator bool,
+	genesisTime time.Time,
+	getTimeHandler func() time.Time,
 ) (*heartbeatMessageInfo, error) {
 
 	if maxDurationPeerUnresponsive == 0 {
 		return nil, ErrInvalidMaxDurationPeerUnresponsive
 	}
+	if getTimeHandler == nil {
+		return nil, ErrNilGetTimeHandler
+	}
 
 	hbmi := &heartbeatMessageInfo{
 		maxDurationPeerUnresponsive: maxDurationPeerUnresponsive,
 		maxInactiveTime:             Duration{0},
 		isActive:                    false,
 		receivedShardID:             uint32(0),
-		timeStamp:                   emptyTimestamp,
-		lastUptimeDowntime:          time.Now(),
+		timeStamp:                   genesisTime,
+		lastUptimeDowntime:          getTimeHandler(),
 		totalUpTime:                 Duration{0},
 		totalDownTime:               Duration{0},
 		versionNumber:               "",
 		nodeDisplayName:             "",
 		isValidator:                 isValidator,
+		genesisTime:                 genesisTime,
+		getTimeHandler:              getTimeHandler,
 	}
-	hbmi.getTimeHandler = hbmi.clockTime
 
 	return hbmi, nil
 }
 
-func (hbmi *heartbeatMessageInfo) clockTime() time.Time {
-	return time.Now()
+func (hbmi *heartbeatMessageInfo) updateFields(crtTime time.Time) {
+	if crtTime.Sub(hbmi.genesisTime) < 0 {
+		return
+	}
+	crtDuration := crtTime.Sub(hbmi.timeStamp)
+	crtDuration = maxDuration(0, crtDuration)
+	previousActive := hbmi.isActive && crtDuration <= hbmi.maxDurationPeerUnresponsive
+	hbmi.isActive = true
+	hbmi.updateMaxInactiveTimeDuration(crtTime)
+	hbmi.updateUpAndDownTime(previousActive, crtTime)
+	hbmi.lastUptimeDowntime = crtTime
 }
 
-func (hbmi *heartbeatMessageInfo) updateFields() {
-	crtDuration := hbmi.getTimeHandler().Sub(hbmi.timeStamp)
+func (hbmi *heartbeatMessageInfo) computeActive(crtTime time.Time) {
+	if crtTime.Sub(hbmi.genesisTime) < 0 {
+		return
+	}
+	crtDuration := crtTime.Sub(hbmi.timeStamp)
 	crtDuration = maxDuration(0, crtDuration)
-
-	hbmi.isActive = crtDuration < hbmi.maxDurationPeerUnresponsive
-	hbmi.updateUpAndDownTime()
-	hbmi.updateMaxInactiveTimeDuration()
+	hbmi.isActive = hbmi.isActive && crtDuration <= hbmi.maxDurationPeerUnresponsive
+	hbmi.updateUpAndDownTime(hbmi.isActive, crtTime)
+	hbmi.lastUptimeDowntime = crtTime
 }
 
 // Wil update the total time a node was up and down
-func (hbmi *heartbeatMessageInfo) updateUpAndDownTime() {
-	lastDuration := hbmi.clockTime().Sub(hbmi.lastUptimeDowntime)
+func (hbmi *heartbeatMessageInfo) updateUpAndDownTime(previousActive bool, crtTime time.Time) {
+	if hbmi.lastUptimeDowntime.Sub(hbmi.genesisTime) < 0 {
```

Comment:
It could be smaller than the genesis Time exactly on the first call of updateUpAndDownTime when the lastUptimeDownTime is not set yet to crtTime. Also crtTime could be before the genesisTime, or?

Reply:
Ok, crtTime can be before genesis time.